apps/arweave/src/ar_nonce_limiter_client.erl

```diff
@@ -0,0 +1,157 @@
+	RawPeers = queue:from_list(Config#config.nonce_limiter_server_trusted_peers),
+	{ok, #state{ remote_servers = RawPeers }}.
+
+handle_call(Request, _From, State) ->
+	?LOG_WARNING("event: unhandled_call, request: ~p", [Request]),
+	{reply, ok, State}.
+
+handle_cast(pull, State) ->
+	ar_util:cast_after(800, ?MODULE, pull),
+	#state{ remote_servers = Q } = State,
+	{{value, RawPeer}, Q2} = queue:out(Q),
+	State2 = State#state{ remote_servers = queue:in(RawPeer, Q2) },
+	case resolve_peer(RawPeer) of
+		error ->
+			%% Push the peer to the back of the queue.
+			{noreply, State2};
+		{ok, Peer} ->
+			case ar_http_iface_client:get_vdf_update(Peer) of
+				{ok, Update} ->
+					#nonce_limiter_update{ session = #vdf_session{
+							step_number = SessionStepNumber } } = Update,
+					case ar_nonce_limiter:apply_external_update(Update, Peer) of
+						ok ->
+							{noreply, State};
+						#nonce_limiter_update_response{ session_found = false } ->
+							case fetch_and_apply_session_and_previous_session(Peer) of
+								ok ->
+									{noreply, State};
+								_ ->
+									{noreply, State2}
+							end;
+						#nonce_limiter_update_response{ step_number = StepNumber }
+								when StepNumber >= SessionStepNumber ->
+							%% We are ahead of this server; try another one, if there are any.
+							{noreply, State2};
+						_ ->
+							case fetch_and_apply_session(Peer) of
+								ok ->
+									{noreply, State};
+								_ ->
+									{noreply, State2}
+							end
+					end;
+				not_found ->
+					%% The server might be restarting. Try another one, if there are any.
+					{noreply, State#state{ remote_servers = queue:in(RawPeer, Q2) }};
+				{error, Reason} ->
+					?LOG_WARNING([{event, failed_to_fetch_vdf_update},
+							{peer, ar_util:format_peer(Peer)},
+							{error, io_lib:format("~p", [Reason])}]),
+					%% Try another server, if there are any.
+					{noreply, State#state{ remote_servers = queue:in(RawPeer, Q2) }}
+			end
+	end;
+
+handle_cast(Cast, State) ->
+	?LOG_WARNING("event: unhandled_cast, cast: ~p", [Cast]),
+	{noreply, State}.
+
+handle_info(Message, State) ->
+	?LOG_WARNING("event: unhandled_info, message: ~p", [Message]),
+	{noreply, State}.
+
+terminate(_Reason, _State) ->
+	ok.
+
+%%%===================================================================
+%%% Private functions.
+%%%===================================================================
+
+resolve_peer(RawPeer) ->
+	case ets:lookup(ar_nonce_limiter, {vdf_server_raw_peer, RawPeer}) of
+		[{_, Peer}] ->
+			{ok, Peer};
+		_ ->
+			case ar_util:safe_parse_peer(RawPeer) of
+				{ok, Peer} ->
+					{ok, Peer};
+				_ ->
+					error
+			end
+	end.
+
+fetch_and_apply_session_and_previous_session(Peer) ->
+	case ar_http_iface_client:get_vdf_session(Peer) of
+		{ok, #nonce_limiter_update{ session = #vdf_session{
+				prev_session_key = PrevSessionKey } } = Update} ->
+			case ar_http_iface_client:get_previous_vdf_session(Peer) of
```

Comment:
I *think* there's a risk of race condition here. If the server switches to a new session between calls to `get_vdf_session` and `get_previous_vdf_session` we may end up getting the same session twice.

Reply:
ah yea, I see that now. I think there's still a chance of missing a session. Like in this case:

1. GET /vdf
- session_key = B
- prev_session_key = A
2. Session changes to C
3. GET /vdf/previous_session
- session_key = B
- prev_session_key = A
4. Hit the retry case since we were expecting session_key = A
5. GET /vdf
 - session_key = C
 - prev_session_key = B
 6. Get /vdf/previous_session
 - session_key = B
 - prev_session_key = A
 7,  ok

In this case we end up with C and B, and never process A (I think). Is that okay?

Reply:
Gotcha! Thanks!